straxen/plugins/acqmon_processing.py

```diff
@@ -0,0 +1,131 @@
+    busy_*  <= V1495 busy veto for tpc channels
+    bb_*    <= V1495 busy veto for high energy tpc channels
+    hev_*   <= DDC10 hardware high energy veto
+    """
+        
+    __version__ ='0.0.19'
+    depends_on = ('raw_records_aqmon') 
+    provides  = ('veto_intervals')
+    data_kind = ('veto_intervals')
+
+
+    aqmon_channel_names = ('sum_wf','m_veto_sync',
+                           'hev_stop', 'hev_start', 
+                           'bb_stop','bb_start',
+                           'busy_stop', 'busy_start')
+    
+    tmp_dtype = strax.time_fields + [(('Interval between start and stop in [ns]', 'interval'), np.int64)]
+
+    def infer_dtype(self):
+        dtype = [(('faux time [ns]', 'time'), np.int64)]
+        dtype += [(('faux endtime [ns]', 'endtime'), np.int64)]
+        for veto in ('busy_','hev_','bb_'):
+            dtype += [(('Start '+ veto +'veto time since unix epoch [ns]', veto + 'time'), np.int64)]
+            dtype += [(('Stop '+ veto +'veto time since unix epoch [ns]', veto + 'endtime'), np.int64)]
+            dtype += [(('Duration of '+ veto +'veto time since unix epoch [ns]', veto + 'interval'), np.int64)]
+        return dtype
+    
+    def setup(self):
+        self.channel_range = self.config['channel_map']['aqmon']
+        self.channel_numbers = np.arange(self.channel_range[1]-7,self.channel_range[1] + 1, 1)
+        self.channel_map = dict(zip(self.aqmon_channel_names,self.channel_numbers))
+        self.t0 = self.config['run_start_time']
+        self.veto_names = ['busy_','bb_','hev_']
+        return self.channel_map
+    
+    def get_window_size(self):
+        # Give a very wide window
+        return (self.config['max_veto_gap']*100)
+    
+    def compute(self, raw_records_aqmon): 
+        rec = strax.raw_to_records(raw_records_aqmon)
+        strax.zero_out_of_bounds(rec)
+        strax.baseline(rec, baseline_samples=10, flip=True)
+        hits = strax.find_hits(rec)
+        
+        vetos = dict()
+        res = dict()
+        for i, veto in enumerate(self.veto_names):
+            channels = channel_select(hits, self.channel_map[veto + 'stop'], self.channel_map[veto + 'start'])
+            vetos[veto + 'veto'] = self.merge_vetos(channels, gap = self.config['max_veto_gap'],\
+                                           dtype = self.tmp_dtype, t = 0)
+        
+        
+        result = strax.dict_to_rec(vetos)
+
+        # Populate fake 'time' and 'endtime' by using the time and endime of the veto singals
+        for name in self.veto_names:
+            if len(result[name + 'veto']['time']) > 0:
+                res = dict(
+                    time = rec['time'][:len(result[name + 'veto']['time'])],
+                    endtime = strax.endtime(rec)[:len(result[name + 'veto']['time'])]) 
+        
+        for v in self.veto_names:
+            for t in ['time', 'endtime', 'interval']:
+                res[v + t] = result[v + 'veto'][t]       
+        
+        return res
```

Comment:
So each line in the return dtype has the generic start and stop times as required by strax, and then also start and stop for each of the 3 veto types (normal busy, HE busy, and HEV). What happens if one chunk has 6 normal busies, 1 HE busy, and 2 HEV? Would the better method be to just have each line be just start, stop, and veto type?

Reply:
I don't remember the exact thing, but as far as I can recall strax didn't like it when I've tried to implement something like that.  What would be the issue with having multiple veto types in one chunk? If the obtained veto interval is !=0 it would populate the resulting dict with an appropriate dtype label. The fake start/end times would be simply derived from the start time of the interval.

Reply:
My question was, what does the output of this plugin look like with this dtype? Can you provide an example?

Reply:
I've just upladed a short the test notebook to our wiki:
https://xe1t-wiki.lngs.infn.it/doku.php?id=xenon:xenon1t:alexelykov:am_veto_interval_notebook

Hope this helps to illustrate things a bit more clearly.

Reply:
This is the cell I was looking for:
```
array([(1595522722021768350, 1595522722021773350, 1595522725885930240, 1595522726048534880, 162604640, 0, 0, 0, 0, 0, 0),
       (1595522722021773350, 1595522722021778350, 1595522726828730740, 1595522726928319730,  99588990, 0, 0, 0, 0, 0, 0),
       (1595522722021778350, 1595522722021783350, 1595522729658690940, 1595522729677668950,  18978010, 0, 0, 0, 0, 0, 0),
       (1595522722021783350, 1595522722021788350, 1595522733431900310, 1595522733514186030,  82285720, 0, 0, 0, 0, 0, 0),
       (1595522722021788350, 1595522722021793350, 1595522736262083060, 1595522736344731590,  82648530, 0, 0, 0, 0, 0, 0)],
      dtype=[(('faux time [ns]', 'time'), '<i8'), (('faux endtime [ns]', 'endtime'), '<i8'), (('Start busy_veto time since unix epoch [ns]', 'busy_time'), '<i8'), (('Stop busy_veto time since unix epoch [ns]', 'busy_endtime'), '<i8'), (('Duration of busy_veto time since unix epoch [ns]', 'busy_interval'), '<i8'), (('Start hev_veto time since unix epoch [ns]', 'hev_time'), '<i8'), (('Stop hev_veto time since unix epoch [ns]', 'hev_endtime'), '<i8'), (('Duration of hev_veto time since unix epoch [ns]', 'hev_interval'), '<i8'), (('Start he_veto time since unix epoch [ns]', 'he_time'), '<i8'), (('Stop he_veto time since unix epoch [ns]', 'he_endtime'), '<i8'), (('Duration of he_veto time since unix epoch [ns]', 'he_interval'), '<i8')])
```
This is storing 11 values per busy/veto interval with >50% zero-padding and unused fields, when we could rather just deal with 3: start, stop, and type. To calculate the total deadtime for a run you would need to do (stop-start) for three separate columns, and this feels really clunky to me.

Reply:
Alright, if I have some spare time I'll look into reducing the dtype to the more essential parts, to steamline this a bit.

Reply:
@darrylmasson 
I redid the veto data kind, to a more streamlined version. Indeed looks better. Is that what you had in mind?

> array([(1595522725885930240, 1595522726048534880, 162604640, 'busy_veto'),
       (1595522726828730740, 1595522726928319730,  99588990, 'busy_veto'),
       (1595522729658690940, 1595522729677668950,  18978010, 'busy_veto'),
       (1595522733431900310, 1595522733514186030,  82285720, 'busy_veto'),
       (1595522736262083060, 1595522736344731590,  82648530, 'busy_veto')],
      dtype=[(('veto start time [ns]', 'time'), '<i8'), (('veto end time [ns]', 'endtime'), '<i8'), (('veto interval [ns]', 'veto_interval'), '<i8'), (('veto signal type', 'veto_type'), '<U10')])

Reply:
Pretty much, yeah. I thought `veto_type` could be a single number (0=busy, 1=HE busy, 2=HEV), but a string is probably more end-user-friendly.